src/blocks/sound/pipewire.rs

```diff
@@ -0,0 +1,643 @@
+
+                        //     for _ in 0..channels {
+                        //         builder_add!(&mut pod_builder, Float(volume)).expect("Unable to push float");
+                        //     }
+
+                        //     unsafe { libspa::pod::builder::Builder::pop(&mut pod_builder, inner_frame.assume_init_mut()) }
+                        // }
+
+                        // unsafe { libspa::pod::builder::Builder::pop(&mut pod_builder, frame.assume_init_mut()) }
+
+                        builder_add!(&mut pod_builder,
+                            Object(
+                                SpaTypes::ObjectParamProps.as_raw(),
+                                ParamType::Props.as_raw()
+                            ) {
+                                SPA_PROP_softVolumes => Array(Float(vec![volume; channels])),
+                                SPA_PROP_channelVolumes => Array(Float(vec![volume; channels])),
+                            }
+                        )
+                        .expect("Unable to add object to pod");
+                        let pod = Pod::from_bytes(&pod_data).expect("Unable to construct pod");
+                        proxy.set_param(ParamType::Props, 0, pod);
+                    }
+                }
+            }
+            Mute(id, mute) => {
+                if let Some(node) = client.data.lock().unwrap().nodes.get(&id) {
+                    if let Some(proxy) = proxies.borrow_mut().proxies_t.get(&node.proxy_id) {
+                        let mut pod_data = Vec::new();
+                        let mut pod_builder = PodBuilder::new(&mut pod_data);
+                        builder_add!(&mut pod_builder,
+                            Object(
+                                SpaTypes::ObjectParamProps.as_raw(),
+                                ParamType::Props.as_raw()
+                            ) {
+                                SPA_PROP_mute => Bool(mute),
+                                SPA_PROP_softMute => Bool(mute),
+                            }
+                        )
+                        .expect("Unable to add object to pod");
+                        let pod = Pod::from_bytes(&pod_data).expect("Unable to construct pod");
+                        proxy.set_param(ParamType::Props, 0, pod);
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct Client {
+    event_listeners: Mutex<Vec<Sender<EventKind>>>,
+    command_listeners: Mutex<Vec<Receiver<CommandKind>>>,
+    pub ready: Mutex<bool>,
+    pub data: Mutex<Data>,
+}
+
+impl Client {
+    fn new() -> Result<Client> {
+        thread::Builder::new()
+            .name("privacy_pipewire".to_string())
+            .spawn(Client::main_loop_thread)
+            .error("failed to spawn a thread")?;
+
+        Ok(Client::default())
+    }
+
+    fn spawn_idle_listener(
```

Comment:
This loop runs when the main loop isn't busy. Not sure if we can spin up an async context instead of looping over the `command_listeners` over and over

Reply:
Does it run right before the main loop goes idle (e.g. before calling `poll()` or equivalent), or does it run repeatedly until main loop has real work to do? If the former, then what wakes the loop when a command is send? If the latter, well, this is bad.

Reply:
It runs repeatedly until the main loop has real work to do, which I agree is bad.